[DR-2773] [DR-2785] Add snapshot endpoints to link, unlink DUOS datasets

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2773 Add new snapshot endpoints
https://broadworkbench.atlassian.net/browse/DR-2785 Expose DUOS link when retrieving snapshot
https://docs.google.com/document/d/11bitEoYHmfb7Z0jQ_H_MFwqj6YCESB8K58LI7l-1dws/edit# original TDR/DUOS integration design

**Background**

This PR adds two new snapshot endpoints to link and unlink DUOS datasets.  These endpoints do the following:
- Verify that the caller holds Sam action `share_policy::reader` on the snapshot (in other words, a caller must be a snapshot steward)
- Verify that the targeted snapshot exists
- Verify that the DUOS ID (if specified) corresponds to an existing dataset
- Executes `SnapshotUpdateDuosDatasetFlight` synchronously and returns the result

**Dev Environment**

My [dev environment](https://jade-ok.datarepo-dev.broadinstitute.org/snapshots) is up to date with my latest changes.

You can set or update a snapshot's existing link to a DUOS dataset via [linkDuosDatasetToSnapshot](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/linkDuosDatasetToSnapshot).

This method expects a DUOS ID corresponding to a valid DUOS dataset.  A few example inputs:
- `DUOS-000003` - maps to a DUOS dev dataset
- `DUOS-000003-unexpected-format` - throws a 400 (unexpected DUOS ID format)
- `DUOS-123456` - throws a 404 (no DUOS dataset found for the DUOS ID)

You can unset a snapshot's existing link to a DUOS dataset via [unlinkDuosDatasetFromSnapshot](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/unlinkDuosDatasetFromSnapshot).

To confirm a snapshot's existing DUOS link, you can [retrieveSnapshot](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/retrieveSnapshot) with the default include list or specifying DUOS as an include value, and examining `duosFirecloudGroup` contents.  This will ultimately be used to expose a snapshot's DUOS link in TDR UI: https://broadworkbench.atlassian.net/browse/DR-2780

You can also observe evidence of a snapshot's DUOS link by looking at its readers: if it's linked to a DUOS dataset, you'll see a Firecloud group email with a name like `DUOS-<nnnnnn>-users@dev.test.firecloud.org` corresponding to its authorized user group.

okotsopoulos committed:
Add link to duos_firecloud_group to snapshot table

okotsopoulos committed:
DuosService and DuosDao refactor

We will take a more atomic approach in our flight to link a snapshot to a DUOS dataset,
so I was able to remove quite a bit of code from DuosService: in most cases our flight can interact directly with DuosDao.
Within DuosDao, made changes to interact with records based on their generated row ID.
Updated unit testing.

okotsopoulos committed:
SnapshotDao changes to retrieve and update a snapshot's DUOS Firecloud group

- When generating a Snapshot from its DB entry, include any associated DuosFirecloudGroupModel
- Added ability to update a snapshot DB entry's pointer to a DUOS Firecloud group entry
- Expanded unit tests

okotsopoulos committed:
Introduce SnapshotUpdateDuosDatasetFlight and steps

The flight can be used for setting, updating, and unsetting a snapshot's link to a DUOS dataset.
I broke the flight down into atomic steps for more straightforward undoing in the event of a flight failure.
The flight response is the DUOS Firecloud groups unlinked from and/or linked to the snapshot.

okotsopoulos committed:
SnapshotDao.updateDuosFirecloudId throws SnapshotUpdateException if no rows updated

Allows for explicit error handling in UpdateSnapshotDuosFirecloudGroupIdStep.
Expanded unit tests to reflect this.
Also ported getDuosFirecloudGroupId to SnapshotDuosFlightUtils and added unit tests for that class.

okotsopoulos committed:
Renaming to clarify when we should perform steps related to Firecloud group creation

okotsopoulos committed:
Move related line of code below its associated comment

okotsopoulos committed:
Add unit tests to verify construction of SnapshotUpdateDuosDatasetFlight

And new test fixtures for creating DuosFirecloudGroupModel objects

okotsopoulos committed:
Add unit tests for SnapshotUpdateDuosDatasetFlight steps, utility methods

okotsopoulos committed:
Merge branch 'develop' into okotsopo/DR-2787-snapshot-link-duos-flight

okotsopoulos committed:
Unit test to confirm that a user can trigger an unlink of a snapshot's DUOS dataset for a snapshot with no existing link

okotsopoulos committed:
Add new IamAction.SHARE_POLICY_READER

If a caller is triggering a link/unlink of a DUOS dataset to/from a snapshot, they must hold this action on the snapshot.
At present, it is conferred by snapshot stewardship only.
This required special handling to account for the presence of colons in the Sam action name, a forbidden character in enum names.
Added unit tests.

okotsopoulos committed:
Introduce DuosClient to interact with consent API

Includes method getDataset which fetches a skeleton DUOS Dataset object for a specified DUOS ID as the logged-in user,
for the time being our only usage is as an existence check.
Also added necessary consent API base path configuration (to be overridden per environment),
DuosDataset record for deserializing our response body,
and unit tests.

okotsopoulos committed:
Introduce SnapshotService.updateSnapshotDuosDataset

The method checks first for the existence of a DUOS dataset if DUOS ID specified.
Only then does it run SnapshotUpdateDuosDatasetFlight synchronously and returns its result.
Added unit tests.

okotsopoulos committed:
Add snapshots API endpoints linkDuosDatasetToSnapshot, unlinkDuosDatasetFromSnapshot

Controller implementations verify that caller holds Sam action 'share_policy::reader' on the snapshot,
since this API call will add and/or remove snapshot readers.

okotsopoulos committed:
[DR-2785] Expose DUOS link by default when retrieving snapshot

While here, elected to fetch snapshot prior to checking for DUOS dataset existence in SnapshotService.updateSnapshotDuosDataset:
if the snapshot doesn't exist, we'll throw before communicating with an external service (DUOS), so will fail faster.
Expanded unit tests.

okotsopoulos commented on src/main/resources/application.properties:
```diff
@@ -124,3 +124,4 @@ usermetrics.metricsReportingPoolSize=10
 usermetrics.syncRefreshIntervalSeconds=3600
 usermetrics.ignorePaths=/api/repository/v1/jobs/*,/api/repository/v1/configs*,/api/repository/v1/upgrade,/actuator/*,/configuration,/status
 rawls.basePath=https://rawls.dsde-dev.broadinstitute.org
+duos.basePath=https://consent.dsde-dev.broadinstitute.org
```
We will need to set `DUOS_BASEPATH` environment variable overrides for the different consent envs:
https://broadworkbench.atlassian.net/browse/DR-2825

pshapiro4broad replied:
I think you'll also need to add a field for DUOS to `ServerSpecification` in clienttests (as we do for `samUri`), assuming that we want to have an integration tests for DUOS.

okotsopoulos replied:
Thanks for the heads up!  For the time being there are no integration tests but this is likely to change.

okotsopoulos commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -1094,6 +1094,110 @@ paths:
+      description: >
+        Link the DUOS dataset to the snapshot so that its authorized users are synced
+        as snapshot readers.
+        NOTE: This is an experimental feature and its response body may change.
```
More context: right now we execute a synchronous flight and return the result, but it's possible we could prefer an asynchronous flight instead.  In that case, we'd return our usual job response which puts the onus on the caller to poll it to completion.

okotsopoulos commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -587,7 +587,7 @@ paths:
           schema:
             type: array
-            default: "SOURCES,TABLES,RELATIONSHIPS,PROFILE,DATA_PROJECT"
+            default: "SOURCES,TABLES,RELATIONSHIPS,PROFILE,DATA_PROJECT,DUOS"
```
Is there a reason we have to set this default value in yaml and redundantly in code?

pshapiro4broad replied:
This is a good question! I think the use of `RETRIEVE_INCLUDE_DEFAULT_VALUE` in the declaration of `SnapshotsApiController.retrieveSnapshot()` is a bug and should be removed:
```
  public ResponseEntity<SnapshotModel> retrieveSnapshot(
      @PathVariable("id") UUID id,
      @Valid
          @RequestParam(
              value = "include",
              required = false,
              defaultValue = RETRIEVE_INCLUDE_DEFAULT_VALUE)
          List<SnapshotRetrieveIncludeModel> include) {
```

It's also used by `SnapshotService.retrieveAvailableSnapshotModel(UUID, AuthenticatedUserRequest)`. I think callers of this method should be changed to use an explicit list of attributes that doesn't have to be kept in sync with the default list. It looks like this code is using the default list as a convenience to avoid specifying a list of attributes in these common cases.

okotsopoulos replied:
Thank you for looking at this more closely!

@nmalfroy, I am vaguely remembering a recent instance where we were tied to some suboptimal default value behavior in our controllers.  Is that ringing a bell?

nmalfroy replied:
Yeah, we shouldn't include access_information by default because it can be really slow (particularly on Azure)

okotsopoulos replied:
Ahh that wasn't quite what I was remembering… it was this conversation, which made me more confident in removing the controller's default parameter configuration:
https://github.com/DataBiosphere/jade-data-repo/pull/1345/files/b6cdb6ffdc84d73f507b767b8c4f7619b213dc6c#r966105813

@nmalfroy I would be interested in chatting more with you about the other changes I made here to remove the redundant default list in code:

https://github.com/DataBiosphere/jade-data-repo/pull/1372/commits/2f524751274f617c237f3ff5b9f0f03158fe89de

I didn't fully remove `SnapshotService.retrieveAvailableSnapshotModel(UUID, AuthenticatedUserRequest)` and force the callers to determine for themselves which fields to include because I wasn't confident about the desired behavior in snapshot export (the only non-test caller).  At any rate, I'm content to have clarified the default behavior within the controller!

okotsopoulos replied:
Flip-flopping back… I reverted this removal.  Our version of `swagger-codegen` 2.4.27 does not support mixing in parameters with refs (I glossed over this in the linked comment above).  So the default is not propagating to the generated API.  We've experimented with `swagger-codegen` upgrades that could fix this, but found that other aspects of code generation broke for us.

For now it felt safest to leave the existing behavior.

okotsopoulos commented on src/main/java/bio/terra/app/controller/SnapshotsApiController.java:
```diff
@@ -65,7 +66,7 @@ public class SnapshotsApiController implements SnapshotsApi {
 
   // We do not include Access_Information since it can get expensive, and for backwards compat
```
Flag for reviewers: re: "backwards compat", do we really not want to introduce new default fields for snapshot retrieval?  I can remove `DUOS` as a default if so.

nmalfroy replied:
Eh, I think it's fine to add fields like this

okotsopoulos commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -0,0 +1,75 @@
+   * @throws DuosDatasetBadRequestException if identifier is malformed
+   * @throws DuosDatasetNotFoundException if no DUOS dataset found
+   */
+  public DuosDataset getDataset(String duosId, AuthenticatedUserRequest userRequest) {
```
Note for reviewers: presently we can interact with DUOS as the authed user, as part of facilitating their interaction with our API.

But when we expand the client to fetch a DUOS dataset's authorized users, we will likely do so as a TDR SA:
- TDR SA is the sole admin of the DUOS dataset's Firecloud user group, and the only user who can alter its members
- DUOS Firecloud group syncing may start out as a process manually triggered by a TDR Admin or similar (and thus will have an authed user request associated), but longer term is likely to run as a scheduled background task

okotsopoulos commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -0,0 +1,75 @@
+              DuosDataset.class);
+      return datasetCall.getBody();
+    } catch (HttpClientErrorException ex) {
+      switch (ex.getStatusCode()) {
```
These are the known possible error conditions per Consent API docs:

https://consent.dsde-dev.broadinstitute.org/#/TDR/get_api_tdr__identifier_

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -1018,6 +1048,9 @@ private SnapshotModel populateSnapshotModelFromSnapshot(
     if (include.contains(SnapshotRetrieveIncludeModel.PROPERTIES)) {
       snapshotModel.properties(snapshot.getProperties());
     }
+    if (include.contains(SnapshotRetrieveIncludeModel.DUOS)) {
```
Note: readers can retrieve a snapshot, and while they might not find information on the snapshot's DUOS link useful, I think that they should be permitted to see it even though they can't set it because DUOS datasets are publicly discoverable to authed users.

However, we may elect not to retrieve this or render it in the TDR UI for snapshot readers.

pshapiro4broad commented on src/main/java/bio/terra/app/configuration/DuosConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties
+@ConfigurationProperties(prefix = "duos")
```
I think you can replace these with a single annotation
```suggestion
@ConfigurationProperties(prefix = "duos")
```

okotsopoulos replied:
I do think we need `@Configuration`, the app fails to start up without it:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

Parameter 0 of constructor in bio.terra.service.duos.DuosClient required a bean of type 'bio.terra.app.configuration.DuosConfiguration' that could not be found.


Action:

Consider defining a bean of type 'bio.terra.app.configuration.DuosConfiguration' in your configuration.
```

But I do think that `@EnableConfigurationProperties` is redundant, I was able to start up the app and see the `duos.basePath` property pulled in.

pshapiro4broad replied:
Interesting, in catalog we don't use it. Maybe that's because our spring app uses `@ConfigurationPropertiesScan`, which scans all classes in a package hierarchy for `@ConfigurationProperties` annotated classes.
https://github.com/DataBiosphere/terra-data-catalog/blob/main/service/src/main/java/bio/terra/catalog/App.java#L28

pshapiro4broad commented on src/main/java/bio/terra/app/configuration/DuosConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+@Configuration
+@EnableConfigurationProperties
+@ConfigurationProperties(prefix = "duos")
+public class DuosConfiguration {
+
+  private String basePath;
+
+  public String getBasePath() {
+    return basePath;
+  }
+
+  public void setBasePath(String basePath) {
+    this.basePath = basePath;
+  }
+}
```
I think this can become a `record`
```suggestion
public record DuosConfiguration(String basePath) {}
```

okotsopoulos replied:
Update, I was not able to get this to work: while reducing boilerplate code is good, I had a hard time getting the immutability of the record to play nicely with Spring Boot.  We may be missing some higher app-level annotation that could enable this (`@ConfigurationPropertiesScan`?) but I'm designating this as OOS.

pshapiro4broad replied:
I was curious about this so I created a branch and (I think) got it working. The one thing I didn't know is that, since TDR is using an older version of Spring Boot than catalog, you need to use the constructor binding annotation for `record`s. See https://github.com/DataBiosphere/jade-data-repo/pull/1374

pshapiro4broad replied:
(This is separate from the `ConfigurationPropertiesScan` change, which lets you remove `Configuration` from this class.)

pshapiro4broad commented on src/main/java/bio/terra/app/controller/SnapshotsApiController.java:
```diff
@@ -313,4 +314,25 @@ public ResponseEntity<List<String>> retrieveUserSnapshotRoles(UUID id) {
+        userReq, IamResourceType.DATASNAPSHOT, id.toString(), IamAction.SHARE_POLICY_READER);
+    SnapshotLinkDuosDatasetResponse response =
+        snapshotService.updateSnapshotDuosDataset(id, userReq, duosId);
+    return new ResponseEntity<>(response, HttpStatus.OK);
```
It's a minor thing but using `ResponseEntity.ok()` is a shorter way to write this. Also personally I would inline the method call 
```suggestion
    return ResponseEntity.ok(snapshotService.updateSnapshotDuosDataset(id, userReq, duosId));
```

okotsopoulos replied:
Good tip, thanks!

pshapiro4broad commented on src/main/java/bio/terra/service/auth/iam/IamAction.java:
```diff
@@ -41,6 +44,10 @@ public enum IamAction {
     this.samActionName = name().toLowerCase();
```
When you have more than one constructer, I think it's usually better to chain constructors when possible (going from ones with fewer arguments to ones with more arguments), as I find it's easier to see how the different constructors relate to each other. In this case:
```suggestion
    this(name().toLowerCase());
```

okotsopoulos replied:
I like this, that makes sense.

okotsopoulos replied:
Actually, this does not work for our enum construction:

```
Cannot reference 'Enum.name' before supertype constructor has been called
```

pshapiro4broad replied:
Oh yeah, was I was typing this I was wondering if that would work, as `enum` construction doesn't allow a few things that you can do elsewhere. Oh well.

pshapiro4broad commented on src/main/java/bio/terra/service/auth/iam/IamAction.java:
```diff
@@ -50,7 +57,7 @@ public String toString() {
   public static IamAction fromValue(String text) {
     for (IamAction b : IamAction.values()) {
-      if (b.name().equals(StringUtils.upperCase(text))) {
+      if (b.samActionName.equalsIgnoreCase(text)) {
```
Super minor comment; since you know `samActionName` is always lowercase you could convert `text` to lowercase outside the `for` and use `equals()` here.

okotsopoulos replied:
That's true.  My reasoning with this change is that it more directly reflects that we only care about a case-insensitive match rather than having to convert one side or the other, but that may boil down to personal preference.

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -0,0 +1,75 @@
+  private final DuosConfiguration duosConfiguration;
+  private final RestTemplate restTemplate;
+
+  @Autowired
```
This annotation is not necessary. Spring assumes that an injectable object's constructor arguments are `Autowired` by default.
```suggestion
```

okotsopoulos replied:
Aha, that rings a bell from your recent Java notes posting!

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosClient.java:
```diff
@@ -0,0 +1,75 @@
+
+  @VisibleForTesting
+  String getDatasetUrl(String duosId) {
+    return String.format("%s/api/tdr/%s", duosConfiguration.getBasePath(), duosId);
```
I guess there's no prebuilt DUOS client library?

okotsopoulos replied:
Correct.  I've flagged our desire for one, but realistically even if they write one it will be awhile before it's made available:

https://broadinstitute.slack.com/archives/CKSCCUZPU/p1667922177959739

okotsopoulos replied:
Greg on DUOS filed this ticket, fingers crossed that auto-generating a client will be an easier-than-expected lift:

https://broadworkbench.atlassian.net/browse/DUOS-2191

pshapiro4broad commented on src/main/java/bio/terra/service/duos/DuosDataset.java:
```diff
@@ -0,0 +1,8 @@
+
+@VisibleForTesting
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record DuosDataset(Integer dataSetId) {}
```
Is `dataSetId` nullable in practice? If not, would use `int` here work instead? That would make it clearer that `null` isn't allowed here.
```suggestion
public record DuosDataset(int dataSetId) {}
```

okotsopoulos replied:
This is a good question!  While in practice we don't expect this to be null I wanted to match this exactly to DUOS' defined model in their code base, since we don't have a client from them:

https://github.com/DataBiosphere/consent/blob/6ec50f6b9ddc1649e17b1f0522107eea3084dddb/src/main/java/org/broadinstitute/consent/http/models/Dataset.java#L12

nmalfroy replied:
I will also say: given that we don't really control it, I always err on the side of boxed objects (I'm allergic to NPEs :) )

pshapiro4broad replied:
Using boxed types doesn't avoid NPEs, it only defers them.

The problem with boxed types is you can end up with NPEs without knowing it, as java won't flag a compile error if you try to use an `Integer` in a math expression, and you still end up with an NPE when the boxed object is `null`. So using boxed types only avoids NPEs if you never use the boxed type as a non-boxed type, or are very careful to check for nulls before using them.

But in this case it's OK since it's a copied type from another API and the code in this PR never looks at the content of this object. If we do ever add code that uses this field, we'll have to be careful.

nmalfroy replied:
IMO it's easier to debug with boxed types.  E.g. if I try to access it, it tells me "boom, I'm null" and points me to the actual problem area.  Not so with the unboxed type.  You end up getting weird stack traces

pshapiro4broad replied:
It's hard to disagree with wanting to avoid weird stack traces...

The benefit I see to not using boxed types is that the compiler will flag an error if you try to assign `null` to an object that can't be null. The IDE can also warn in a case where a method that it knows can return `null` is used. So this turns what would be a runtime error into a compile time error.

But this only works if a field can never be `null`. If it can possibly be null, then you have to use a boxed type. If it can't possibly be null, then used an unboxed type. If it's somewhere in between -- it should never be `null`, but we can't rule it out completely -- then you have to decide which way you want to find out about the error.

pshapiro4broad commented on src/test/java/bio/terra/service/auth/iam/IamActionTest.java:
```diff
@@ -0,0 +1,35 @@
+
+@ActiveProfiles({"google", "unittest"})
+@Category(Unit.class)
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
```
These tests don't appear to use mocking so I don't think this annotation is needed.

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -0,0 +1,112 @@
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+
+@RunWith(SpringRunner.class)
```
Does this test need this annotation? It looks like it's a pure unit test -- no injections or non-mocked classes are used. If so, you don't need to initialize Spring and you could use
```
@RunWith(MockitoJUnitRunner.StrictStubs.class)
```
or
```
@ExtendWith(MockitoExtension.class)
```
And use `@Mock` below instead of `@MockBean`. I like `ExtendWith` better but it's Junit5 and I'm not sure how much of TDR is using that yet.

nmalfroy replied:
I do think that you might need it in order to use the Mocks below

pshapiro4broad replied:
`SpringRunner` is only necessary if you need a spring container for injections. In general, a "real" unit test doesn't use injections at all and only uses mocks. Using `MockitoJUnitRunner` gives you support for `@Mock` but doesn't have the overhead of initializing a spring container. It also lets you use strict stubbing which makes tests more precise.

The cases where you need a spring container are for MVC tests (which should be using `WebMvcTest` anyway, not `SpringRunner`) and DAO tests where you need a live database connection.

okotsopoulos replied:
Thanks for this extra explanation, it's helped me understand where we'd use one over the other!

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -0,0 +1,112 @@
+    userReq =
+        AuthenticatedUserRequest.builder()
+            .setSubjectId("DuosClientTest")
+            .setEmail("DuosClientTest@unit.com")
```
These fields are unused and can be left blank (`""`).

pshapiro4broad commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -0,0 +1,112 @@
+        AuthenticatedUserRequest.builder()
+            .setSubjectId("DuosClientTest")
+            .setEmail("DuosClientTest@unit.com")
+            .setToken("token")
```
This field is used, but no test asserts that it's being used property. Maybe it's too hard to test this with the way it's used?

okotsopoulos replied:
I had to play with this a bit, but ended up with a custom `ArgumentMatcher` to verify that the `HttpEntity` had the token set as its sole means of authentication.  This may be overkill in other similar tests but as we're apt to call the consent API as both an authenticated user and a TDR SA, I think it's worth it here.

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotServiceTest.java:
```diff
@@ -140,11 +156,7 @@ public void testRetrieveSnapshot() {
     assertThat(
         service.retrieveAvailableSnapshotModel(snapshotId, TEST_USER),
         equalTo(
-            new SnapshotModel()
```
It's always nice to see common code being refactored :)

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotServiceTest.java:
```diff
@@ -125,12 +137,16 @@ public class SnapshotServiceTest {
+  private final UUID duosFirecloudGroupId = UUID.randomUUID();
   private final Instant createdDate = Instant.now();
+  private final DuosFirecloudGroupModel duosFirecloudGroup =
+      DuosFixtures.duosFirecloudGroupFromDb(DUOS_ID, duosFirecloudGroupId);
```
Why is this method called `FromDb()`? Does it make a database call to create this object? Since `SnapshotDao` is mocked I wouldn't expect this test to be using the database.

okotsopoulos replied:
Ahhh, I can see how this naming is confusing.  This fixture returns a `DuosFirecloudGroupModel` populated in a manner matching the guarantees of a record retrieved from the DB (ex. has id populated), but it doesn't make a DB call to get it.

Would `mockDuosFirecloudGroupFromDb` be clearer, or do you have another naming suggestion to disambiguate?  (I am not interacting with mocks in this call either, so wasn't sure if that presented a different point of confusion.)

pshapiro4broad replied:
Ok, I see that you need both a `db` version and a non-`db` one. I think this is fine as long as it's clear to the people working on the code. Maybe a javadoc in the text fixtures class would help? I think part of my confusion was that the fixtures classes I've worked with before in in TDR are concerned with creating "live" objects, not ones for mock tests, so I assumed that this was also a live database call.

Looking at `DuosFixtures`, I would rename `duosFirecloudGroupCreated()` to `createFirecloudGroup()`. I'd put the action first, and remove `duos` since it's in the class name already. I would rename `duosFirecloudGroupFromDb()` to `createDbFirecloudGroup()`, since `db` is a modifier of `FirecloudGroup`. But names are very subjective so whatever names seem best to you are OK with me!

okotsopoulos replied:
Good call, I've added javadocs to these methods. We have a mix of live resource creation and mocked object creation in fixtures, it would be nice to have clearer delineation between what you're getting.  Thanks for the naming thoughts, I've been staring at this for too long!

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotServiceTest.java:
```diff
@@ -781,4 +797,109 @@ public void verifySnapshotInaccessibleBySamFallsBackToPassport() {
+    assertThat("Job is submitted and response returned", result, equalTo(jobResponse));
+
+    // Verify that we do not try to check for DUOS dataset existence given an unspecified DUOS ID
+    verify(duosClient, never()).getDataset(any(), eq(TEST_USER));
```
It would probably be a lot of work, but if this test class used strict stubs then you wouldn't need to verify `never()` as that would be the default for all mocks. I wonder if you could enable strict stubs on a per-test basis to migrate the code...

pshapiro4broad reviewed (approved):
Looks OK to me as far as I can tell, I have a few comments

nmalfroy commented on src/test/java/bio/terra/service/duos/DuosClientTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    when(duosConfiguration.getBasePath()).thenReturn(BASE_PATH);
+
+    userReq =
```
FWIW, we often have a static final with the default user to use.  I actually wonder if we should have a fixture that just has a common static user so we don't need to keep redefining it

okotsopoulos replied:
I had this thought too!

pshapiro4broad replied:
At some point TDR should update to the latest version of terra-common-lib, which deprecates `AuthenticatedUserRequest` and replaces it with a wrapper around a bearer token string. It will be a big change since this object is used everywhere but it helps in cases like this; only the bearer token is used so it's unnecessary to set email and subject ID.

okotsopoulos replied:
Until we migrate to latest `terra-common-lib` I introduced static method `AuthenticationFixtures.randomUserRequest()` for reuse.

pshapiro4broad replied:
Unless you need multiple different users for a single test, you could also create a static `user` object instead of making a new one for each test.

nmalfroy reviewed (approved):
this looks great, thanks!

okotsopoulos committed:
Merge branch 'develop' into okotsopo/DR-2773-snapshot-link-duos-endpoints

okotsopoulos committed:
Correct bug in merge conflict resolution - duplicate method

okotsopoulos committed:
@EnableConfigurationProperties is redundant

okotsopoulos committed:
Use ResponseEntity.ok to streamline SnapshotsApiController returns

okotsopoulos committed:
@Autowired no longer required on @Component constructors (or constructors for extending annotations)

okotsopoulos committed:
Inline method calls when constructing DUOS link responses in SnapshotsApiController

okotsopoulos committed:
Removed unused code

Removed instances of journalService in dataset, snapshot controllers
Snapshot retrieval now uses OpenAPI default includes value
We still have a variant of SnapshotService.retrieveAvailableSnapshotModel which tries to mimic the default behavior of the API:
outside of tests, it's only called when preparing the manifest in a SnapshotExportFlight and I wasn't sure exactly what snapshot information was desired there.
For now, I consolidated this default code within the method / removed from the controller to disincentivize its use here, but ideally the flight should establish what it would like retrieved.

okotsopoulos committed:
Unit test streamlining

Add AuthenticationFixtures.randomUserRequest() to generate test user
Only use @SpringRunner and @MockBean where needed to test non-mocked injected components: this saves on unnecessary overhead in test startup
Move DuosFixtures initialization to static variables rather than redefining on setup for each test
Remove times(1) in calls to verify: this is the default number of method calls verified

okotsopoulos committed:
Favor assertThat over assertEquals - I've been reversing my expected and actual arguments in assertEquals

okotsopoulos committed:
Refactor DuosFixtures method names - thanks @pshapiro4broad

okotsopoulos committed:
Revert removal of SnapshotsApiController.RETRIEVE_INCLUDE_DEFAULT_VALUE

We set the default value of retrieveSnapshot endpoint's includes value in our OpenAPI spec, but our version of swagger-codegen 2.4.27 does not pull it into the generated SnapshotsApi.
This is because we define includes via a reference to a separately defined schema.
We've tried to upgrade to a 3.x version of swagger-codegen in the hopes of fixing this, but this broke other aspects of our code generation.
For now, we need to maintain a default value in the controller.
